Sources/NIOCore/IPAddresses.swift

```diff
@@ -0,0 +1,447 @@
+    /// - returns: The `IPAddress` for the given string or `nil` if the string representation is not supported.
+    public init<Bytes: Collection>(packedBytes bytes: Bytes) throws where Bytes.Element == UInt8, Bytes.Index == Int {
+        switch bytes.count {
+        case 4: self = .v4(.init(packedBytes: bytes))
+        case 16: self = .v6(.init(packedBytes: bytes))
+        default:
+            throw IPAddressError.bytesArrayHasWrongLength(bytes.count)
+        }
+    }
+    
+    /// Creates a new `IPAddress` for the given libc IPv4 address.
+    ///
+    /// - parameters:
+    ///     - posixIPv4Address: libc ipv4 address.
+    public init(posixIPv4Address: in_addr) {
+        let uint8Bitmask: UInt32 = 0x000000FF
+        
+        let uint8AddressBytes: IPv4Bytes = .init((
+            UInt8((posixIPv4Address.s_addr >> 24) & uint8Bitmask),
+            UInt8((posixIPv4Address.s_addr >> 16) & uint8Bitmask),
+            UInt8((posixIPv4Address.s_addr >> 8) & uint8Bitmask),
+            UInt8(posixIPv4Address.s_addr & uint8Bitmask)
+        ))
+        
+        self = .v4(.init(address: uint8AddressBytes))
+    }
+    
+    /// Creates a new `IPAddress` for the given libc IPv6 address.
+    ///
+    /// - parameters:
+    ///     - posixIPv6Address: libc ipv6 address.
+    public init(posixIPv6Address: in6_addr) {
+        self = .v6(.init(address: .init(posixIPv6Address.__u6_addr.__u6_addr8)))
+    }
+}
+
+extension UInt8 {
+    /// Convenience function especially useful for representing IPv6 bytes as readable string.
+    func toHex() -> Character {
+        switch self {
+        case 0: return "0"
+        case 1: return "1"
+        case 2: return "2"
+        case 3: return "3"
+        case 4: return "4"
+        case 5: return "5"
+        case 6: return "6"
+        case 7: return "7"
+        case 8: return "8"
+        case 9: return "9"
+        case 10: return "A"
+        case 11: return "B"
+        case 12: return "C"
+        case 13: return "D"
+        case 14: return "E"
+        case 15: return "F"
+        default: preconditionFailure()
+        }
+    }
+}
+
+/// We define an extension on `IPv4Bytes` that gives it a collection conformance.
+extension IPv4Bytes: Collection {
+    public typealias Index = Int
+    public typealias Element = UInt8
+    
+    public var startIndex: Index { 0 }
+    public var endIndex: Index { 4 }
+    
+    public subscript(position: Index) -> Element {
+        get {
+            switch position {
+            case 0: return self.bytes.0
+            case 1: return self.bytes.1
+            case 2: return self.bytes.2
+            case 3: return self.bytes.3
+            default: preconditionFailure()
+            }
+        }
+    }
+    
+    public func index(after: Index) -> Index {
+        return after + 1
+    }
+}
+
+/// We define an extension on `IPv6Bytes` that gives it a collection conformance.
+extension IPv6Bytes: Collection {
```

Comment:
Same here, this should conform to `MutableCollection` and `RandomAccessCollection`.